### PR TITLE
Present forms in embedded

### PR DESCRIPTION
**Pull request by porter-stripe**
## Summary
- Now when tapping a selection in embedded, if the PM requires input we display a form to collect input and use that as the selected payment option.
- Generate a new form upon update to preserve selections
- Handle edge cases around canceling and reselecting the old option

## Motivation
- Embedded

## Testing
- Manual
- Existing tests
- New UI tests

## Changelog
N/A

### CODE REVIEW

**Comment by porter-stripe** [comment]
Added skip dead code check as all the flags are for unused protocol function parameters for functions that will be implementing in following PRs.

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedFormViewController.swift [hunk 1]
```diff
@@ -70,25 +70,29 @@ class EmbeddedFormViewController: UIViewController {
...
         return paymentMethodFormViewController.paymentOption
     }
-
-    private let loadResult: PaymentSheetLoader.LoadResult
+    
     private let paymentMethodType: PaymentSheet.PaymentMethodType
     private let configuration: EmbeddedPaymentElement.Configuration
     private let intent: Intent
     private let elementsSession: STPElementsSession
+    private let savedPaymentMethods: [STPPaymentMethod]?
```
Comment (wooj-stripe) [inline StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedFormViewController.swift:89]: 1. Seems like `savedPaymentMethods` can be non-optional
2. We technically don't even have to pass in the entire array. Instead we can either pass in:

* a boolean -- "shouldUseNewCardNewCardHeader: Bool" or
* `savedPaymentMethod.first` type: `STPPaymentMethod?`

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedPaymentElement+Internal.swift [hunk 2]
```diff
@@ -90,8 +92,55 @@ extension EmbeddedPaymentElement: EmbeddedPaymentMethodsViewDelegate {
         delegate?.embeddedPaymentElementDidUpdateHeight(embeddedPaymentElement: self)
     }
 
-    func selectionDidUpdate() {
-        delegate?.embeddedPaymentElementDidUpdatePaymentOption(embeddedPaymentElement: self)
+    func selectionWasTapped(didUpdate: Bool) {
+        // Defer notifying delegate until we are about to exit this function to give time to either allow the new payment option to come from
+        // The new `EmbeddedFormViewController` instance if needed
```
Comment (wooj-stripe) [inline StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedPaymentElement+Internal.swift:97]: Saying "to give time to either allow the new.." does not sound precise.  How about:
```
Deferring notifying delegate until the exit of this function guarantees the new payment option comes from the new instance of `EmbeddedFormViewController`
```

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedPaymentElement+Internal.swift [hunk 3]
```diff
@@ -90,8 +92,55 @@ extension EmbeddedPaymentElement: EmbeddedPaymentMethodsViewDelegate {
...
+            }
+        }
+        
+        guard case let .new(paymentMethodType) = embeddedPaymentMethodsView.selection else {
+            // This can occur when selection is being reset to nothing selected, so don't assert.
+            return
+        }
+
+        guard let presentingVC = presentingViewController else {
+            assertionFailure("Presenting view controller not found, set EmbeddedPaymentElement.presentingViewController.")
```
Comment (wooj-stripe) [inline StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedPaymentElement+Internal.swift:110]: stpAssertionFailure("")

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedPaymentElement+Internal.swift [hunk 4]
```diff
@@ -90,8 +92,55 @@ extension EmbeddedPaymentElement: EmbeddedPaymentMethodsViewDelegate {
...
+                delegate?.embeddedPaymentElementDidUpdatePaymentOption(embeddedPaymentElement: self)
+            }
+        }
+        
+        guard case let .new(paymentMethodType) = embeddedPaymentMethodsView.selection else {
+            // This can occur when selection is being reset to nothing selected, so don't assert.
+            return
+        }
+
+        guard let presentingVC = presentingViewController else {
```
Comment (wooj-stripe) [inline StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedPaymentElement+Internal.swift:109]: nit:
        guard let presentingViewController else {

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedPaymentElement+Internal.swift [hunk 5]
```diff
@@ -90,8 +92,55 @@ extension EmbeddedPaymentElement: EmbeddedPaymentMethodsViewDelegate {
...
+            paymentMethodType: paymentMethodType,
+            previousPaymentOption: self.formViewController?.previousPaymentOption,
+            analyticsHelper: analyticsHelper,
+            formCache: formCache
+        )
+        formViewController.delegate = self
+
+        // Only show forms that require user input
+        guard formViewController.collectsUserInput else {
+            self.formViewController = nil  // Clear out any previous form view controller to update self._paymentOption
```
Comment (wooj-stripe) [inline StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedPaymentElement+Internal.swift:128]: 1. s/self._paymentOption/self.paymentOption ??
2. I don't understand this comment -- I would have expected to see a didSet for `formViewController` to clear out `paymentOption`

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedPaymentElement+Internal.swift [hunk 6]
```diff
@@ -90,8 +92,55 @@ extension EmbeddedPaymentElement: EmbeddedPaymentMethodsViewDelegate {
...
+        }
+
+        let bottomSheet = BottomSheetViewController(
+            contentViewController: formViewController,
+            appearance: configuration.appearance,
+            isTestMode: configuration.apiClient.isTestmode,
+            didCancelNative3DS2: {
+                stpAssertionFailure("3DS2 was triggered unexpectedly")
+            }
+        )
```
Comment (wooj-stripe) [inline StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedPaymentElement+Internal.swift:139]: nit: since i'm seeing duplicate code in 3 spots now... maybe should add a function:
```
    func bottomSheetController(with viewController: BottomSheetContentViewController) -> BottomSheetViewController {
        return BottomSheetViewController(contentViewController: viewController,
                                         appearance: configuration.appearance,
                                         isTestMode: configuration.apiClient.isTestmode,
                                         didCancelNative3DS2: {
            stpAssertionFailure("3DS2 was triggered unexpectedly")
        })
    }
```

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedPaymentElement.swift [hunk 7]
```diff
@@ -127,15 +127,30 @@ public final class EmbeddedPaymentElement {
...
+            var formViewController: EmbeddedFormViewController?
+            if let formPaymentMethodType = self?.formViewController?.selectedPaymentOption?.paymentMethodType {
+                formViewController = EmbeddedFormViewController(configuration: configuration,
+                                                                intent: loadResult.intent,
+                                                                elementsSession: loadResult.elementsSession,
+                                                                savedPaymentMethods: loadResult.savedPaymentMethods,
+                                                                paymentMethodType: formPaymentMethodType,
+                                                                previousPaymentOption: self?.formViewController?.selectedPaymentOption,
+                                                                analyticsHelper: analyticsHelper,
+                                                                formCache: .init()) // Always use a fresh cache after an update
```
Comment (wooj-stripe) [inline StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedPaymentElement.swift:141]: nit: instead of having formCache a required arg, make the default parameter use a fresh cache?

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedPaymentElement.swift [hunk 8]
```diff
@@ -127,15 +127,30 @@ public final class EmbeddedPaymentElement {
...
+                
+            }
+            
+            // 2.4 Re-initialize embedded view to update the UI to match the newly loaded data.
             let embeddedPaymentMethodsView = Self.makeView(
                 configuration: configuration,
                 loadResult: loadResult,
                 analyticsHelper: analyticsHelper,
-                previousPaymentOption: self?._paymentOption,
+                previousPaymentOption: formViewController != nil ? formViewController?.selectedPaymentOption : self?._paymentOption, // Use the possibly new payment option from the formVC if it exists, self?._paymentOption is not yet updated with the new formVC
```
Comment (wooj-stripe) [inline StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedPaymentElement.swift:150]: 1. we should break this into a separate line
2. I'm not following what we're doing here, because the computed variable has:
```
        if let formPaymentOption = formViewController?.selectedPaymentOption {
            return formPaymentOption
        }
```
I'm sure i'm missing something, but it seems like we could just pass in self?._paymentOption without checking formViewController here.

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedPaymentElement.swift [hunk 9]
```diff
@@ -155,6 +170,8 @@ public final class EmbeddedPaymentElement {
             self.intent = loadResult.intent
             self.embeddedPaymentMethodsView = embeddedPaymentMethodsView
             self.containerView.updateEmbeddedPaymentMethodsView(embeddedPaymentMethodsView)
+            self.formViewController = formViewController
+            self.formCache = .init()
```
Comment (wooj-stripe) [inline StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedPaymentElement.swift:174]: I'm not following why this class needs a formCache. Seems like only `formViewController` would need a copy since we aren't preserving the cache between update calls?

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedPaymentMethodsView.swift [hunk 10]
```diff
@@ -24,12 +24,20 @@ class EmbeddedPaymentMethodsView: UIView {
...
     private(set) var selection: Selection? {
+        willSet {
+            previousSelection = selection
+        }
         didSet {
             updateMandate()
-            if oldValue != selection {
-                delegate?.selectionDidUpdate()
-            }
+            delegate?.selectionWasTapped(didUpdate: oldValue != selection)
```
Comment (wooj-stripe) [inline StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedPaymentMethodsView.swift:40]: I think there are times where the selectionDidUpdate, but a user didn't actually tap (e.g. removal use case), so I think the name selectionDidUpdate() is more appropriate.

**Review by wooj-stripe — COMMENTED** [review]
Leaving comments a bit early, didn't review test code yet.

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedPaymentElement.swift (hunk 9 above)
Comment (porter-stripe) [reply StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedPaymentElement.swift:174]: The cache needs to be kept in storage between creations (presentations) of formViewController, to preserve input of multiple PMs, and partial user input. That way if we
1. Present Klarna enter email
2. Close Klarna form
3. Open card, start typing a number
4. Close card
5. Open Klarna, we still have the cached input in Klarna email field.

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedFormViewController.swift (hunk 1 above)
Comment (porter-stripe) [reply StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedFormViewController.swift:89]: Ah, I like this better

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedPaymentElement.swift (hunk 8 above)
Comment (porter-stripe) [reply StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedPaymentElement.swift:149]: 1. Broke this into it's own line
2. paymentOption looks at `self.formViewController`, but at this point in the function `self.formViewController` is not yet updated with the latest instance, `self.formViewController` is updated with this new instance later in the function with the rest of the properties. So we need to look at the freshly created formViewController if it exists for a previous payment option. I added a better comment here.

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedPaymentMethodsView.swift (hunk 10 above)
Comment (porter-stripe) [reply StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedPaymentMethodsView.swift]: Updated!

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedPaymentMethodsView.swift [hunk 11]
```diff
@@ -189,9 +197,20 @@ class EmbeddedPaymentMethodsView: UIView {
             delegate?.heightDidChange()
         }
     }
+    
+    // MARK: Internal functions
+    func resetSelectionToLastSelection() {
```
Comment (wooj-stripe) [inline StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedPaymentMethodsView.swift:202]: 1. `resetSelectionToLastSelection` looks super similar to `didTap` -- maybe we find a way to share some code?
2.  I am seeing a case where i instantiate embedded, saved card is selected, i click on "new card" then tap cancel, and my original selection is not preserved.

Comment (porter-stripe) [reply StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedPaymentMethodsView.swift:212]: Refactored #1 and fixed #2.

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedPaymentMethodsView.swift [hunk 12]
```diff
@@ -114,6 +129,7 @@ class EmbeddedPaymentMethodsView: UIView {
             if initialSelection == selection {
                 applePayRowButton.isSelected = true
                 self.selection = initialSelection
+                self.selectedRowButton = savedPaymentMethodButton
```
Comment (wooj-stripe) [inline StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedPaymentMethodsView.swift:132]: I'm having trouble following this logic.. here's how i'm reading it:
1.  If the initial selection is apple pay, then..
  i. applePayRowButton should be selected
  ii. set the self.selection = initialSelection
  iii. set the selectedRowButton back to the savedPaymentMethodButton (but wait, in "i", we've set apple pay row button to be selected?)

File: Example/PaymentSheet Example/PaymentSheetUITest/EmbeddedUITest.swift [hunk 13]
```diff
@@ -14,11 +14,43 @@ class EmbeddedUITests: PaymentSheetUITestCase {
...
         app.buttons["Present embedded payment element"].waitForExistenceAndTap()
-        // TODO: Test card form (see PaymentSheetVerticalUITests testUpdate)
-
-        // Selecting Alipay w/ deferred PaymentIntent...
+        
+        // Entering a card w/ deferred PaymentIntent...
+        app.buttons["Card"].waitForExistenceAndTap()
+        XCTAssertTrue(app.staticTexts["Add card"].waitForExistence(timeout: 10))
+        try! fillCardData(app, postalEnabled: true)
+        app.toolbars.buttons["Done"].tap()
```
Comment (wooj-stripe) [inline Example/PaymentSheet Example/PaymentSheetUITest/EmbeddedUITest.swift:25]: 1. waitForExistenceAndTap() is probably more reliable.
2. TIL: didn't realize there was toolbars.buttons, i've always used "app.buttons["Done"]"

File: Example/PaymentSheet Example/PaymentSheetUITest/EmbeddedUITest.swift [hunk 14]
```diff
@@ -14,11 +14,43 @@ class EmbeddedUITests: PaymentSheetUITestCase {
...
+        XCTAssertTrue(app.buttons["Continue"].isEnabled)
+        app.buttons["Continue"].waitForExistenceAndTap()
+        XCTAssertTrue(app.staticTexts["Payment method"].waitForExistence(timeout: 10))
+        XCTAssertEqual(app.staticTexts["Payment method"].label, "•••• 4242")
+        
+        // ...and *updating* to a SetupIntent...
+        app.buttons.matching(identifier: "Setup").element(boundBy: 1).tap()
```
Comment (wooj-stripe) [inline Example/PaymentSheet Example/PaymentSheetUITest/EmbeddedUITest.swift:32]: waitForExistenceAndTap() is probably more reliable.

File: Example/PaymentSheet Example/PaymentSheetUITest/EmbeddedUITest.swift [hunk 15]
```diff
@@ -14,11 +14,43 @@ class EmbeddedUITests: PaymentSheetUITestCase {
...
+        // ...(wait for it to finish updating)...
+        _ = app.buttons["Reload"].waitForExistence(timeout: 10)
```
Comment (wooj-stripe) [inline Example/PaymentSheet Example/PaymentSheetUITest/EmbeddedUITest.swift:34]: XCTAssertTrue on waitForExistence to get a better fail state in our tests. -- this should be applied to any time we "Reload"

File: Example/PaymentSheet Example/PaymentSheetUITest/EmbeddedUITest.swift [hunk 16]
```diff
@@ -14,11 +14,43 @@ class EmbeddedUITests: PaymentSheetUITestCase {
...
+        
+        // ....Tapping card should show the card form with details preserved
+        app.buttons["Card"].waitForExistenceAndTap()
+        // ...thus the Continue button should be enabled
+        app.buttons["Continue"].waitForExistenceAndTap()
+        // ...should cause the card ending in 4242 that was previously entered to be the selected payment method
+        XCTAssertTrue(app.staticTexts["Payment method"].waitForExistence(timeout: 10))
+        XCTAssertEqual(app.staticTexts["Payment method"].label, "•••• 4242")
+        // ...switching from setup to payment should preserve this card as the selected payment method
+        app.buttons.matching(identifier: "Payment").element(boundBy: 1).tap()
```
Comment (wooj-stripe) [inline Example/PaymentSheet Example/PaymentSheetUITest/EmbeddedUITest.swift:46]: s/tap/waitForExistenceAndTap

File: Example/PaymentSheet Example/PaymentSheetUITest/EmbeddedUITest.swift [hunk 17]
```diff
@@ -27,6 +59,49 @@ class EmbeddedUITests: PaymentSheetUITestCase {
...
+        // ...go back into deferred PaymentIntent mode
+        app.buttons.matching(identifier: "Payment").element(boundBy: 1).tap()
+        // ...(wait for it to finish updating)...
+        _ = app.buttons["Reload"].waitForExistence(timeout: 10)
+        //...selecting Klarna w/ deferred PaymentIntent...
+        app.buttons["Klarna"].waitForExistenceAndTap()
+        // ...fill out the form for Klarna
+        let emailField = app.textFields["Email"]
+        XCTAssertTrue(emailField.waitForExistence(timeout: 10))
+        emailField.tap()
```
Comment (wooj-stripe) [inline Example/PaymentSheet Example/PaymentSheetUITest/EmbeddedUITest.swift:86]: s/tap/waitForExistenceAndTap

File: Example/PaymentSheet Example/PaymentSheetUITest/EmbeddedUITest.swift [hunk 18]
```diff
@@ -293,6 +368,110 @@ class EmbeddedUITests: PaymentSheetUITestCase {
...
+        settings.mode = .paymentWithSetup
+        settings.uiStyle = .paymentSheet
+        settings.customerKeyType = .legacy
+        settings.formSheetAction = .continue
+        settings.customerMode = .new
+        loadPlayground(app, settings)
+        
+        // Start by saving a new card
+        app.buttons["Present PaymentSheet"].waitForExistenceAndTap()
+        try! fillCardData(app, postalEnabled: true)
```
Comment (wooj-stripe) [inline Example/PaymentSheet Example/PaymentSheetUITest/EmbeddedUITest.swift:383]: I think we can use a .returning customer here so we don't have to fillCardData?

File: StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedPaymentMethodsView.swift (hunk 12 above)
Comment (porter-stripe) [reply StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/Embedded/EmbeddedPaymentMethodsView.swift:132]: Good point, the applePayRowButton.isSelected = true is a bit repetitive and reads weird. I moved this below as just one line for all.

File: Example/PaymentSheet Example/PaymentSheetUITest/EmbeddedUITest.swift (hunk 18 above)
Comment (porter-stripe) [reply Example/PaymentSheet Example/PaymentSheetUITest/EmbeddedUITest.swift:383]: Yeah I'd like to, but the returning customer has two PMs saved and the one that is selected by default changes randomly between a card and bank making it not great for tests.

File: Example/PaymentSheet Example/PaymentSheetUITest/EmbeddedUITest.swift [hunk 19]
```diff
@@ -293,6 +367,110 @@ class EmbeddedUITests: PaymentSheetUITestCase {
...
     }
+    
+    func testSelection() {
+        var settings = PaymentSheetTestPlaygroundSettings.defaultValues()
+        settings.mode = .paymentWithSetup
+        settings.uiStyle = .paymentSheet
+        settings.customerKeyType = .legacy
+        settings.formSheetAction = .continue
+        settings.customerMode = .new
+        loadPlayground(app, settings)
```
Comment (wooj-stripe) [inline Example/PaymentSheet Example/PaymentSheetUITest/EmbeddedUITest.swift:378]: We should add a test, which tests that selection disappears when a payment method disappears.
e.g.
1. Have mobilepay selected w/ payment mode.
2. Switch to payment+sfu (which forces an update)

Ensure that mobile pay is not the selected payment method, and it is not showing on screen.

This is important because the init code for `EmbeddedPaymentMethodsView` is a bit complex -- it's easy to think  we can just set `self.selection = initialSelection` prior to calling `super.init(frame: .zero)` but this would result in showing that the unavailable payment method is still available (despite it not being selected)

**Review by wooj-stripe — APPROVED** [review]
